### PR TITLE
Vet clinic database: query and update animals table

### PR DESCRIPTION
**Pull request by zhorabay**
- [ ] In this branch, I:
- Added a column **_species_** of type _string_ to the **_animals_** table. Modified the **_schema.sql_** file.
- Inserted the provided data and added them to the data.sql file.
- Inside a transaction updated the animals table by setting the species column to unspecified. Verified that change was made.
Screen Shot 2023-10-24 at 20 26 14

- Then rolled back the change and verified that the species columns went back to the state before the transaction.
Screen Shot 2023-10-24 at 17 34 39

- Updated the animals table by setting the species column to digimon for all animals that have a name ending in mon.
- Updated the animals table by setting the species column to pokemon for all animals that don't have species already set.
- Committed the transaction and verified that changes persist after commit.
Screen Shot 2023-10-24 at 17 39 09

- Inside a transaction deleted all records in the animals table
Screen Shot 2023-10-24 at 17 42 25

- Then rolled back the changes
Screen Shot 2023-10-24 at 17 42 51

- Deleted all animals born after Jan 1st, 2022.
Screen Shot 2023-10-24 at 17 47 29

- Created a savepoint for the transaction.
- Updated all animals' weight to be their weight multiplied by -1.
Screen Shot 2023-10-24 at 17 49 05

- Rollbacked to the savepoint
- Updated all animals' weights that are negative to be their weight multiplied by -1 and committed
Screen Shot 2023-10-24 at 17 50 45

- Added all these transactions to the queries.sql file.
- Wrote queries to answer the following questions:
- How many animals are there?
Screen Shot 2023-10-24 at 18 02 21

- How many animals have never tried to escape?
Screen Shot 2023-10-24 at 18 09 17

- What is the average weight of animals?
Screen Shot 2023-10-24 at 18 10 47

- Who escapes the most, neutered or not neutered animals?
Screen Shot 2023-10-24 at 18 14 11

- What is the minimum and maximum weight of each type of animal?
Screen Shot 2023-10-24 at 20 30 12

- What is the average number of escape attempts per animal type of those born between 1990 and 2000?
Screen Shot 2023-10-24 at 21 25 28

- Added these queries to the queries.sql file.

### CODE REVIEW

File: queries.sql [hunk 1]
```diff
@@ -8,3 +8,65 @@ SELECT (name, escape_attempts) FROM animals WHERE weight_kg > 10.50;
 SELECT * FROM animals WHERE neutered = true;
 SELECT * FROM animals WHERE name != 'Gabumon';
 SELECT * FROM animals WHERE weight_kg BETWEEN 10.40 AND 17.30;
+
+BEGIN;
+UPDATE animals
+SET species = NULL;
+SELECT * FROM animals;
+ROLLBACK;
+SELECT * FROM animals;
```
Comment (shahadat3669) [inline queries.sql]: 👍 You're on the right track, but there's a minor adjustment needed in your SQL script to fully meet the project requirements:

- [ ] The provided SQL script updates the "species" column of the "animals" table to set all its values to NULL, which is great. However, as per the project requirement, you should first update the "species" column to set it to 'unspecified' within a transaction, verify that the change was made, and then roll back the change to verify that the "species" column returns to its original state before the transaction. Please update your query accordingly to align with the project requirements.
- [ ] If needed, kindly update the screenshot in the pull request (PR) description. Your cooperation is greatly appreciated.

Your dedication to following project guidelines is much appreciated! Keep it up! 👏📜🔍

File: queries.sql [hunk 2]
```diff
@@ -8,3 +8,65 @@ SELECT (name, escape_attempts) FROM animals WHERE weight_kg > 10.50;
...
+SELECT AVG(weight_kg) FROM animals;
+SELECT neutered, COUNT(*)
+FROM animals
+WHERE escape_attempts > 0
+GROUP BY neutered;
+SELECT neutered,
+       MIN(weight_kg),
+       MAX(weight_kg)
+FROM animals
+GROUP BY neutered;
```
Comment (shahadat3669) [inline queries.sql]: 👍 Great work on the SQL query! It's well-structured, and you've provided useful insights. However, to meet the project requirements, you should make the following adjustments:

- [ ] Your SQL query currently groups data from the "animals" table by the "neutered" column, calculating the minimum (MIN) and maximum (MAX) values of the "weight_kg" column for each group. This approach offers insights into the range of weights for neutered and non-neutered animals. However, as per the project requirement, your query should answer the question: "What is the minimum and maximum weight of each type of animal?" To meet this requirement, please update your query to calculate the minimum and maximum weight for each distinct animal type within the dataset, providing a comprehensive summary of the weight range for different animal types.
- [ ] If needed, kindly update the screenshot in the pull request (PR) description. Your cooperation is greatly appreciated.

Your dedication to aligning your work with the project's needs is highly appreciated. Keep up the good work! 👏📊🐕

File: queries.sql [hunk 3]
```diff
@@ -8,3 +8,65 @@ SELECT (name, escape_attempts) FROM animals WHERE weight_kg > 10.50;
...
+SELECT neutered,
+       AVG(escape_attempts)
+FROM animals
+WHERE date_of_birth >= '1990-01-01' AND date_of_birth <= '2000-12-31'
+GROUP BY neutered;
```
Comment (shahadat3669) [inline queries.sql]: 👍 You've got a strong SQL query there, but to fully meet the project requirements, you should make a slight adjustment:

- [ ] The SQL query you've provided retrieves data from the "animals" table, focusing on the "neutered" column and calculating the average number of "escape_attempts" for animals born between January 1, 1990, and December 31, 2000, grouped by the "neutered" column. This query provides valuable insights into the relationship between neutering status and escape behavior for animals within that date range. However, as per the project requirement, you should specifically answer the question: "What is the average number of escape attempts per animal type of those born between 1990 and 2000?" To meet this requirement, please update your query to provide the average number of escape attempts per distinct animal type (digimon, pokemon) for animals born between 1990 and 2000.
- [ ] If needed, kindly update the screenshot in the pull request (PR) description. Your cooperation is greatly appreciated.

Your commitment to aligning with the project requirements is greatly appreciated! Keep up the good work! 👏📊🐾

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi @zhorabay, 

Great job so far! 👏 You've done well, and you're almost there for the next project ⏭️, but there are a few issues 🐛 that you still need to work on.




## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ put correct sql commands in the correct files
✅ add the required datas in the data.sql file


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadata3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: queries.sql [hunk 4]
```diff
@@ -8,3 +8,63 @@ SELECT (name, escape_attempts) FROM animals WHERE weight_kg > 10.50;
...
+WHERE escape_attempts > 0
+GROUP BY neutered;
+SELECT
+       MIN(weight_kg),
+       MAX(weight_kg)
+FROM animals
+SELECT
+       AVG(escape_attempts)
+FROM animals
+WHERE date_of_birth >= '1990-01-01' AND date_of_birth <= '2000-12-31'
```
Comment (addod19) [inline queries.sql]: - [ ] Here is how we can improve the query and output the right query results we are expection
```
   SELECT species, AVG(escape_attempts) FROM animals
   WHERE  date_of_birth >= '1990/01/01' AND date_of_birth <= '2000/12/31'
   GROUP BY species;

```
We are expected to answer the question below
Screenshot from 2023-10-24 18-10-13

**Review by addod19 — CHANGES_REQUESTED** [review]
## Hi @zhorabay, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
- [ ] Great job making some of the changes :rocket: 


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: queries.sql (hunk 4 above)
Comment (Abenezer-Tilahun) [reply queries.sql]: Fixed.

**Review by Abenezer-Tilahun — APPROVED** [review]
Hi @zhorabay 
 
Your project is complete  👍 💯   There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 👏🏻

✔️ No linter errors.
✔️ You used correct Git-Flow.
✔️ Professional README  
✔️ Good Commit Messages
✔️ Good job for Implemented Suggested changes.


## Optional suggestions 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

**Pull request merged**